### PR TITLE
Adding information about RESTDataSource to the migration page and fix highlighting 

### PR DESCRIPTION
**Pull request by rkoron007**
Don't merge this until https://github.com/apollographql/apollo-server/pull/6949 is merged in!

### CODE REVIEW

**Comment by rkoron007** [comment]
Don't merge this until https://github.com/apollographql/apollo-server/pull/6949 is merged in!

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:401]: It's correct as is, but typically only use template strings if we need interpolation as well via `${}`
```suggestion
      'movies',
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:469]: As above
```suggestion
      'movies',
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:484]: stray or intentional?

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:590]: I'm not sure what problem the `TDataSources` generic is solving - seems like hard-typing that `Record<string, DataSource>` everywhere is sufficient, and I'm not sure why (or even how?) I might extend that if I was using this snippet.

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:609]: Not sure the `buildServer` abstraction is helpful here, it's just hiding how AS is constructed.

```suggestion
const server = new ApolloServer({
  schema,
  plugins: [ApolloDataSources({ dataSources })],
});
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:604]: 👍

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:590]: I'm not too sure about the answer here, @smyrick would you happen to know?

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:484]: stray, I'll remove it!
```suggestion
```

File: docs/source/migration.mdx
Comment (flyboarder) [reply docs/source/migration.mdx:589]: it's required for the DataSourcesFn, otherwise `options.dataSources()` returns an `Any` type.

Comment (trevor-scheer) [reply docs/source/migration.mdx:589]: I see the usefulness of the `DataSourcesFn`, but I guess I don't see why it can't just be:

```ts
type DataSourcesFn =  () => DataSources;
```

or in other words, as the consumer of these types, why would I ever need to plug in my own type for `TDataSources` rather than rely on the default which is the `DataSources` record?

Comment (flyboarder) [reply docs/source/migration.mdx:589]: Sure I agree, I don't think there is really any added benefit to the optional type parameter.

@rkoron007 could probably simplify the whole example to just...
```
import { ApolloServerPlugin, BaseContext } from '@apollo/server';
import type { DataSource } from 'apollo-datasource';

type DataSources = Record<string, DataSource>;
type DataSourcesFn = () => DataSources;

interface ContextWithDataSources extends BaseContext {
  dataSources?: DataSources
}

export const ApolloDataSources = (options: {dataSources: DataSourcesFn}):ApolloServerPlugin<ContextWithDataSources> => ({
  requestDidStart: async (requestContext) => {
    const dataSources = options.dataSources();
    const initializers = Object.values(dataSources).map(async (dataSource) => {
      if (dataSource.initialize) dataSource.initialize({cache: requestContext.cache, context: requestContext.contextValue})
    });

    await Promise.all(initializers);  
    
    requestContext.contextValue.dataSources = dataSources;
  }
});
```

Comment (rkoron007) [reply docs/source/migration.mdx:589]: Okay, I'll do that! Thank you both so much 🚀

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:418]: What's up with all the highlight changes? Is `highlight-line` buggy?

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:508]:
```suggestion
> In Apollo Server 4, we've moved `apollo-datasource-rest` to the new [`@apollo/datasource-rest`](https://www.npmjs.com/package/@apollo/datasource-rest) package. Most of the functionality between the two packages is the same. However, some small syntax differences exist in how we pass a request's `headers`, `params`, `cacheOptions`, and `body`. See [Fetching from REST](./data/fetching-rest#method-parameters) for more details.
```

maybe? i just don't know that "updated X to the new Y package" really reads right.

Alternatively "we've updated X and moved it to the new Y package"

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:602]: I do wonder if this will fully work end-to-end if you also add a `context` function, since the type returned by it won't match what's expected in resolvers... or is that "fixed" by the `dataSources` being optional?  might help to extend the example to include a `context` function so folks know what to return from it (and maybe even a resolver that reads context).

**Review by glasser — COMMENTED** [review]
otherwise looking good though

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:418]: Yeah, the translation between TS -> JS with these longer lines can sometimes create a new highlighted line depending on the line length. From the authoring standpoint, I honestly think it looks cleaner using `highlight-start` and `end` on these long lines. Let me know if you disagree; this is my personal preference and by no means our standard!

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:602]: So this might not be the correct way to pass in the context's type, but I took a stab at making a complete example:
```ts
import { ApolloServerPlugin, BaseContext } from "@apollo/server";
import type { DataSource } from "apollo-datasource";
import UserAPI from "./datasources/users";

type DataSources = Record<string, DataSource>;
type DataSourcesFn = () => DataSources;

interface ContextWithDataSources extends BaseContext {
  dataSources?: DataSources;
}

export const ApolloDataSources = (options: {
  dataSources: DataSourcesFn,
}): ApolloServerPlugin<ContextWithDataSources> => ({
  requestDidStart: async (requestContext) => {
    const dataSources = options.dataSources();
    const initializers = Object.values(dataSources).map(async (dataSource) => {
      if (dataSource.initialize)
        dataSource.initialize({
          cache: requestContext.cache,
          context: requestContext.contextValue,
        });
    });

    await Promise.all(initializers);

    requestContext.contextValue.dataSources = dataSources;
  },
});

interface MyContext {
  token?: string
  dataSources?: {
    userAPI: UserAPI
  }
}

const server = new ApolloServer<MyContext>({
  typeDefs,
  resolvers: {
    Query: {
      books: (_, __, { dataSources, token }) => {
        // Our contextValue has access to our dataSources 
        // and any *other* values we set in our context function 
        return dataSources.userAPI.getUser(token); // highlight-line
      },
    }
  },
  plugins: [ApolloDataSources({
    dataSources: () => {
      return { userAPI: new UserAPI() } // highlight-line
    }
  })]
});

const { url } = await startStandaloneServer(server, {
  async context({ req }) {
    return { token: req.headers.authorization || "" }
  },
});

console.log(`🚀 Server listening at: ${url}`);
```

File: docs/source/migration.mdx
Comment (glasser) [reply docs/source/migration.mdx:418]: Ah cool. So we should generally avoid `highlight-line` on MultiCodeBlock?  Does DXE know about this, maybe it's fixable?

File: docs/source/migration.mdx
Comment (glasser) [reply docs/source/migration.mdx:602]: One minor note here: this only works if the tsconfig settings `strictNullChecks` and `strict` are at their admittedly default settings of `false`; it's pretty common to set them to true these days!  With one of them set to true, this won't work:

```
      me: (_, __, { dataSources, token }) => {
        // Our contextValue has access to our dataSources 
        // and any *other* values we set in our context function 
        return dataSources.userAPI.getUser(token); // highlight-line
      },
```

because TypeScript will (correctly!) realize that `dataSources` and `token` might not be defined, and so you'll need to do `dataSources!.userAPI.getUser(token!)` instead.

Comment (glasser) [reply docs/source/migration.mdx:602]: Given that you're not actually showing this in the docs I'll approve, but just be aware of that if you do end up filling out the example somewhere...

**Review by glasser — APPROVED** [review]

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:602]: This is so good to know, thank you for this context! I super appreciate knowing about these TS best practices for the future! 🙇

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:418]: I'd say in 90% of cases `highlight-line` works great; it's just in long lines where it can be a problem.

Comment (rkoron007) [reply docs/source/migration.mdx:418]: [Actually, I saw Daniel push a fix for this!](https://github.com/trevorblades/remark-typescript/commit/db4bf5619d3cd07aad759a5270ecd083a640fbf8). I just merged this, and I don't think it's worth reverting this one, but good to know for the future!

**Pull request merged**
